### PR TITLE
PXC-2999 : DML commands are not replicating to PXC-8.0 nodes through …

### PR DESCRIPTION
**Pull request by kennt-percona**
…MySQL shell

Issue
DML commands (coming from mysqlsh) are not replicating to the other nodes
This is because the SQL sessions/threads used by the mysqlsh are not
wsrep enabled.

Solution
Setup the threads with wsrep.

### CODE REVIEW

File: mysql-test/suite/galera/include/have_mysqlx_plugin.inc [hunk 1]
```diff
@@ -0,0 +1,9 @@
+if (`SELECT count(plugin_status)=0 FROM information_schema.plugins WHERE plugin_name LIKE 'mysqlx'`)
```
Comment (dutow) [inline mysql-test/suite/galera/include/have_mysqlx_plugin.inc]: This file already exists in `mysql-test/include`

File: mysql-test/suite/galera/include/xplugin_create_user.inc [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#==== Usage ====
```
Comment (dutow) [inline mysql-test/suite/galera/include/xplugin_create_user.inc]: This also exists in `include/`

File: mysql-test/suite/galera/include/xplugin_drop_user.inc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+--disable_query_log
```
Comment (dutow) [inline mysql-test/suite/galera/include/xplugin_drop_user.inc]: This also exists in include/

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+--echo # node 1
+
+--write_file $MYSQL_TMP_DIR/pxc_mysqlx_script.tmp
+-->sql
+ DROP DATABASE if EXISTS dbtest;
+ CREATE DATABASE dbtest;
+ USE dbtest;
+
+ CREATE TABLE t1 (
+  id int PRIMARY KEY,
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: INT

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+  f2 text);
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: TEXT

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+# Setup the test database and test simple CRUD operations
+# and see if they replicate
+--echo ## Testing CREATE TABLE DDL and INSERT DML (SQL)
+--connection node_1
+--echo # node 1
+
+--write_file $MYSQL_TMP_DIR/pxc_mysqlx_script.tmp
+-->sql
+ DROP DATABASE if EXISTS dbtest;
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: IF

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+ INSERT INTO t1 VALUES (2, "bar");
+-->endsql
+EOF
+
+--exec $MYSQLXTEST -u x_root --password='' --file=$MYSQL_TMP_DIR/pxc_mysqlx_script.tmp 2>&1
+--remove_file $MYSQL_TMP_DIR/pxc_mysqlx_script.tmp
+
+--connection node_1
+--echo # node 1
+select * from dbtest.t1;
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: SELECT... and so on, use upper case for SQL everywhere

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+--let $wait_condition = SELECT COUNT(*) = 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = 'dbtest' AND TABLE_NAME = 't1'
+--source include/wait_condition.inc
+
+--let $wait_condition = SELECT COUNT(*) = 2 FROM dbtest.t1
+--source include/wait_condition.inc
+
+select * from dbtest.t1;
+
+#
+# Now run the delete test
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: These comments above the echos are redundant

File: mysql-test/suite/galera/t/pxc_mysqlx.test [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+--connection node_2
+--echo # node 2
+--let $wait_condition = SELECT COUNT(*) = 1 FROM dbtest.t1
+--source include/wait_condition.inc
+
+select * from dbtest.t1;
+
+
+# Now run the update case
+--echo ## Testing UPDATE DML (SQL)
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_mysqlx.test]: Double #

File: sql/srv_session.cc [hunk 10]
```diff
@@ -792,6 +796,7 @@ bool Srv_session::open() {
   DBUG_ASSERT(state == SRV_SESSION_CREATED || state == SRV_SESSION_CLOSED);
 
   thd.push_protocol(&protocol_error);
+
```
Comment (dutow) [inline sql/srv_session.cc]: accidental change?

File: sql/srv_session.cc [hunk 11]
```diff
@@ -1107,8 +1129,16 @@ int Srv_session::execute_command(enum enum_server_command command,
...
+#ifdef WITH_WSREP
+  if (thd.wsrep_cs().state() == wsrep::client_state::s_exec) {
+    wsrep_after_command_before_result(&thd);
+  }
+  wsrep_after_command_after_result(&thd);
+#endif /* WITH_WSREP */
+
   thd.pop_protocol();
   DBUG_ASSERT(thd.get_protocol() == &protocol_error);
+
```
Comment (dutow) [inline sql/srv_session.cc]: Unnecessary change

File: sql/srv_session.cc [hunk 12]
```diff
@@ -841,6 +846,14 @@ bool Srv_session::open() {
     return true;
   }
 
+#ifdef WITH_WSREP
+  if (thd.wsrep_cs().state() != wsrep::client_state::s_none) {
+    thd.wsrep_cs().cleanup();
+  }
+  thd.wsrep_client_thread = true;
```
Comment (dutow) [inline sql/srv_session.cc:853]: Shouldn't we lock LOCK_wsrep_thd before changing this? wsrep_sync_wait depends on it

**Review by dutow — CHANGES_REQUESTED** [review]
Could we also add a case for a failing DDL/DML (to ensure that the server doesn't crash / work correctly)?

File: sql/srv_session.cc [hunk 13]
```diff
@@ -1095,6 +1113,10 @@ int Srv_session::execute_command(enum enum_server_command command,
 
   mysql_audit_release(&thd);
 
+#ifdef WITH_WSREP
+  wsrep_before_command(&thd);
```
Comment (kamil-holubicki) [inline sql/srv_session.cc:1120]: If we look at "normal" flow, in sql_parse.cc do_command() which is caller of dispatch_command() (so is equivalent of this function) we call wsrep_before_command() as well, but we have more sophisticated logic around this which detects BF aborted transactions done by background rollbacker.
There is also logic that blocks queries in case of snapshot was not installed.

Don't we need this logic here as well?

Comment (kennt-percona) [reply sql/srv_session.cc:1120]: I followed the async slave (rpl) implementation for these calls, which doesn't implement the retry logic at this level.  It's probably too complicated to add at this point.

File: sql/srv_session.cc (hunk 12 above)
Comment (dutow) [reply sql/srv_session.cc:853]: Question not answered.

File: sql/srv_session.cc (hunk 11 above)
Comment (dutow) [reply sql/srv_session.cc]: Still there.

File: sql/srv_session.cc (hunk 10 above)
Comment (dutow) [reply sql/srv_session.cc]: Still there.

File: mysql-test/suite/galera/t/pxc_mysqlx.test (hunk 9 above)
Comment (dutow) [reply mysql-test/suite/galera/t/pxc_mysqlx.test]: Still there.

File: mysql-test/suite/galera/t/pxc_mysqlx.test (hunk 7 above)
Comment (dutow) [reply mysql-test/suite/galera/t/pxc_mysqlx.test]: Please change the  keywords in the test file to upper case.

**Comment by dutow** [comment]
"Could we also add a case for a failing DDL/DML (to ensure that the server doesn't crash / work correctly)?" - I also don't see a testcase for this?

**Comment by dutow** [comment]
@kennt-percona As I said on the standup, we should fix the above comments. But as these are all about tetst (except the locking question), we can also create a separate jira ticket about fixing them and merge this so we don't block the GA build with test fixes.

File: mysql-test/suite/galera/t/pxc_mysqlx.test (hunk 9 above)
Comment (kennt-percona) [reply mysql-test/suite/galera/t/pxc_mysqlx.test]: oops, forgot to push my changes

Comment (dutow) [reply mysql-test/suite/galera/t/pxc_mysqlx.test]: This one is still there, other test changes look good now.

**Review by dutow — APPROVED** [review]
Please fix that ##, otherwise the tests look good now.
